File: s05_multiple_linear_regression.py
# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('data/s5_50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding categorical data [§5 Lect41: "MLR - Python pt1" ...@6min50]
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder = LabelEncoder()
X[:, 3] = labelencoder.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap [§5 Lect41: "MLR - Python pt1" ...@12min20]
# [...but @13min20 says LinearRegression library takes care of it, so not needed (but can run if want to)]
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set [NB. `sklearn.cross_validation` is deprecated]
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting Multiple Linear Regression to the Training set [§5 Lect42: "MLR - Python pt2"]
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results [§5 Lect43: "MLR - Python pt3"]
y_pred = regressor.predict(X_test)

# Building the optimal model using Backward Elimination [§5 Lect44: "MLR - Python (Backward Elimination): Prep"]
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)
# cont'd [§5 Lect45: "MLR - Python (Backward Elimination): HOMEWORK !"]
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit() # "intercept [b0] is not included by default and should be added"
regressor_OLS.summary()
X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# cont'd [§5 Lect46: "MLR - Python (Backward Elimination): Homework solution"]
X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
X_opt = X[:, [0, 3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

File: s05_multiple_linear_regression.R
# Multiple Linear Regression

# Importing the dataset
dataset = read.csv('data/s5_50_Startups.csv')
# dataset = dataset[, 2:3]

# Encoding categorical data [§5 Lect48: "MLR - R pt1" ...@3min00]
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Multiple Linear Regression to the Training set [§5 Lect49: "MLR - R pt2"]
regressor = lm(formula = Profit ~ .,
               data = training_set)

# Predicting the Test set results [§5 Lect50: "MLR - R pt3"]
y_pred = predict(regressor, newdata = test_set)

# Building the optimal model using Backward Elimination [§5 Lect51: "MLR - R (Backward Elimination): HOMEWORK !"]
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)
# Optional Step: Remove State2 only (as opposed to removing State directly)
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + factor(State, exclude = 2),
#                data = dataset)
# summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)
# cont'd [§5 Lect52: "MLR - R (Backward Elimination): Homework solution"]
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

